_transcripts/033-rust-1.50-1.51.md

```diff
@@ -0,0 +1,1232 @@
+sort of magical language-provided things, to more like library type, almost.
+Still a primitive, still fundamental, and privileged, too, with their own
+syntax, but way less of their own, like, special little snowflake sort of data
+structure.
+
+__Jon__: Yeah, it's really nice.
+
+The next one is kind of interesting because this ties back to I think what we
+talked about in 1.49 in the previous episode.
+
+__Ben__: Certainly, yeah.
+
+__Jon__: Which was— in 1.49, just to sort of recap, was, it suddenly enabled you
+to implement `Drop` for unions, and it allowed union fields to not just have to
+be `Copy`, but they could also be `ManuallyDrop`. And the idea here is that when
+you drop a union, you don't know which variant the union is. That's sort of the
+point of a union. It might be one of many different things, so you can't
+actually drop any of the fields. And any type that implements `Copy` does not
+implement `Drop`. So that's why a union could implement `Drop` and contain
+`Copy` types. And what changed in 1.49 was that you— it was also allowed to
+implement `Drop` if it contained `ManuallyDrop` types, because they don't get
+dropped, they're manually dropped. And what was realized in 1.50 was that
+because they are `ManuallyDrop`, it should be safe to assign to them as well.
+Right? So imagine that you have a union. You don't really know which variant you
+have of the union. It's always safe to just assign to a given field of any of
+the variants of the union if it's `ManuallyDrop`, because the act of writing to
+the union is safe, right? Because reading from a union is unsafe anyway. And the
+value you overwrite gets dropped, but it implements `ManuallyDrop` so it doesn't
+get dropped. So the change here is that now it is safe to assign to a
+`ManuallyDrop` union field, whereas previously it was—
+
+__Ben__: Yeah, this was a surprise.
+
+__Jon__: Yeah, it was only safe to assign to `Copy` fields before.
+
+__Ben__: Because my mental model of unions is like, well, they're the unsafe
+version of enums. And so, like, everything is unsafe, in my mind, for a union.
+But actually, it makes sense, sort of like how making a raw pointer in Rust
+isn't unsafe, although we'll get to that in a sec. But reading it is. In this
+way, sort of, writing to a union is totally fine in this case. But reading from
+it isn't.
+
+__Jon__: Yeah.
+
+__Ben__: So Rust is pretty precise about this, to not have to have too much
+unsafe lying around, where it's not needed.
+
+__Jon__: And I do really like this sort of endeavor to find— let's make all the
+things that are safe be safe. Another thing that that I thought was— like, it
+seems like a sort of small change. But it gets us a really cool optimization
+that I don't think we've talked about on the podcast before. And this is the
+idea that on Unix platforms, `File` now has a niche. *ˈnich*? *ˈnēsh*?
+
+__Ben__: I say *ˈnēsh*. I'm not sure; I'm not an English (_unintelligible—
+9:57_).
```

Comment:
I think maybe I just don't know what word is being used here after "English".

Reply:
I honestly have no idea. It sounds like "English-andrean", but that makes no sense :sweat_smile: